controllers/servingruntime_controller.go

```diff
@@ -404,57 +406,55 @@ func runtimeSupportsPredictor(rt *kserveapi.ServingRuntimeSpec, p *api.Predictor
 	return true
 }
 
-// getRuntimesSupportingPredictor returns a list of keys for runtimes that support the predictor p
+// getRuntimesSupportingPredictor returns a map of keys for runtimes that support the predictor p
 //
 // A predictor may be supported by multiple runtimes.
-func (r *ServingRuntimeReconciler) getRuntimesSupportingPredictor(ctx context.Context, p *api.Predictor) ([]types.NamespacedName, error) {
+func (r *ServingRuntimeReconciler) getRuntimesSupportingPredictor(ctx context.Context, p *api.Predictor) (map[string]types.NamespacedName, error) {
 	// list all runtimes
 	runtimes := &kserveapi.ServingRuntimeList{}
 	if err := r.Client.List(ctx, runtimes, client.InNamespace(p.Namespace)); err != nil {
 		return nil, err
 	}
 
-	// list all cluster serving runtimes
-	csrs := &kserveapi.ClusterServingRuntimeList{}
-	if r.HasNamespaceAccess {
-		if err := r.Client.List(ctx, csrs); err != nil {
-			return nil, err
-		}
-	}
-
 	restProxyEnabled := r.ConfigProvider.GetConfig().RESTProxy.Enabled
-	srnns := make([]types.NamespacedName, 0, len(csrs.Items)+len(runtimes.Items))
-	rtNames := make(map[string]struct{})
+	srnns := make(map[string]types.NamespacedName)
+
 	for i := range runtimes.Items {
 		rt := &runtimes.Items[i]
 		if rt.Spec.IsMultiModelRuntime() && runtimeSupportsPredictor(&rt.Spec, p, restProxyEnabled, rt.Name) {
-			rtNames[rt.Name] = struct{}{}
-			srnns = append(srnns, types.NamespacedName{
+			srnns[rt.Name] = types.NamespacedName{
 				Name:      rt.Name,
 				Namespace: p.Namespace,
-			})
+			}
 		}
 	}
-	for i := range csrs.Items {
-		crt := &csrs.Items[i]
-		if crt.Spec.IsMultiModelRuntime() && runtimeSupportsPredictor(&crt.Spec, p, restProxyEnabled, crt.Name) {
-			if _, ok := rtNames[crt.Name]; !ok {
-				srnns = append(srnns, types.NamespacedName{
-					Name:      crt.Name,
-					Namespace: p.Namespace,
-				})
+
+	// list all cluster serving runtimes
+	if r.EnableCSRWatch {
+		csrs := &kserveapi.ClusterServingRuntimeList{}
+		if err := r.Client.List(ctx, csrs); err != nil {
+			return nil, err
+		}
+		for i := range csrs.Items {
+			crt := &csrs.Items[i]
+			if crt.Spec.IsMultiModelRuntime() && runtimeSupportsPredictor(&crt.Spec, p, restProxyEnabled, crt.Name) {
+				if _, ok := srnns[crt.Name]; !ok {
+					srnns[crt.Name] = types.NamespacedName{
+						Name:      crt.Name,
+						Namespace: p.Namespace,
+					}
+				}
```

Comment:
If you move the CSR part here before the regular SRs, this can just be a map addition, no need to have the existence check first.

Reply:
I guess you would rather have more map additions/updates then checks for existence. Sure, I can make that change.